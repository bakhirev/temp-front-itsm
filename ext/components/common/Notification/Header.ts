import styled from 'styled-components';

import { Subtitle3 } from '../../Typography';

export const Header = styled(Subtitle3)<{ hasOffset?: boolean }>`
  ${({ hasOffset }) => (hasOffset ? 'margin-bottom: 4px;' : '')}
`;
