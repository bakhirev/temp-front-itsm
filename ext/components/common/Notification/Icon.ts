import styled from 'styled-components';

export const Icon = styled.div`
  position: absolute;
  left: 12px;
  top: 12px;
`;
