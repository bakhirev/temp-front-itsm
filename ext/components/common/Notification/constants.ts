export const TRANSFORM_SCALE = 0.9;
export const DEFAULT_DURATION = 500;

export const ICON_CLOSE_RIGHT = 16;
export const ICON_CLOSE_TOP = 12;

export const STATIC_PADDING = '12px 16px 12px 52px';
export const DYNAMIC_PADDING = '14px 52px';
export const MOBILE_DYNAMIC_PADDING = '14px 24px';

export const MOBILE_MARGIN = '16px';
export const TOP_MARGIN = '16px';
