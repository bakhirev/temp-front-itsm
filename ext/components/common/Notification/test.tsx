import React, { FC } from 'react';
import { shallow } from 'enzyme';

import type { INotificationProps } from './NotificationComponent';
import type { IDynamicNotificationProps } from './DynamicNotification';

export const runCommonTests = (Notification: FC<INotificationProps>, caption: string) => {
  const header = caption + ' Header';
  const message = caption + ' Message';
  return [
    it('should render component by default as "INFO" notification without header', () => {
      const wrapper = shallow(<Notification>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component by default as "INFO" notification with header', () => {
      const wrapper = shallow(<Notification header={header}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="INFO" notification without header', () => {
      const wrapper = shallow(<Notification kind={'info'}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="INFO" notification with header', () => {
      const wrapper = shallow(
        <Notification kind={'info'} header={header}>
          {message}
        </Notification>
      );
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="WARNING" notification without header', () => {
      const wrapper = shallow(<Notification kind={'warning'}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="WARNING" notification with header', () => {
      const wrapper = shallow(
        <Notification kind={'warning'} header={header}>
          {message}
        </Notification>
      );
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="SUCCESS" notification without header', () => {
      const wrapper = shallow(<Notification kind={'success'}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="SUCCESS" notification with header', () => {
      const wrapper = shallow(
        <Notification kind={'success'} header={header}>
          {message}
        </Notification>
      );
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="ERROR" notification without header', () => {
      const wrapper = shallow(<Notification kind={'error'}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="ERROR" notification with header', () => {
      const wrapper = shallow(
        <Notification kind={'error'} header={header}>
          {message}
        </Notification>
      );
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component by default as "INFO" notification without header with custom width', () => {
      const wrapper = shallow(<Notification width={777}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),
  ];
};

export const runCommonDynamicTests = (
  Notification: FC<IDynamicNotificationProps>,
  caption: string
) => {
  const message = caption + ' Message';
  return [
    it('should render component inside the document.body container', () => {
      const wrapper = shallow(<Notification container={document.body}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),
  ];
};
