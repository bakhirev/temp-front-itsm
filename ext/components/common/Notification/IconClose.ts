import styled from 'styled-components';

import { DEFAULT_THEME } from '../default-theme';
import { markerStyle } from '../marker';

import { ICON_CLOSE_RIGHT, ICON_CLOSE_TOP } from './constants';

export const IconClose = styled.div<{ markerOffset: number; inverse?: boolean }>`
  position: absolute;
  right: ${ICON_CLOSE_RIGHT}px;
  top: ${ICON_CLOSE_TOP}px;
  display: flex;
  justify-content: center;
  align-items: center;
  &:hover {
    cursor: pointer;
  }
  ${({ theme, markerOffset, inverse }) => markerStyle(theme, markerOffset, inverse)}
`;

IconClose.defaultProps = {
  theme: DEFAULT_THEME,
};
