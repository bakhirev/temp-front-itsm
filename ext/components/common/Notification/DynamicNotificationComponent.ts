import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../default-theme';
import { MEDIA_MOBILE } from '../media';
import { Z_INDEXES } from '../z-indexes';

import { NotificationComponent } from './NotificationComponent';
import { Header } from './Header';
import { Message } from './Message';
import { Icon } from './Icon';
import { IconClose } from './IconClose';
import { getInverseColor } from './utils';
import { DYNAMIC_PADDING, MOBILE_DYNAMIC_PADDING, MOBILE_MARGIN, TOP_MARGIN } from './constants';
import type { INotificationComponentProps } from './NotificationComponent';

export interface IDynamicNotificationComponentProps extends INotificationComponentProps {
  inverse?: boolean;
  mobile?: boolean;
  managed?: boolean;
  top?: number;
  left?: number;
  right?: number;
}

export const DynamicNotificationComponent = styled(
  NotificationComponent
)<IDynamicNotificationComponentProps>`
  padding: ${DYNAMIC_PADDING};

  ${({ top, left, right, managed }) =>
    managed
      ? `
    ${top !== undefined ? `margin-top: ${top}px;` : ''}
    ${left !== undefined ? `margin-left: ${left}px;` : ''}
    ${
      right !== undefined
        ? `
      margin-left: auto;
      margin-right: ${right}px;
    `
        : ''
    }
    &:not(:first-child) {
      margin-top: ${TOP_MARGIN};
    }
    @media ${MEDIA_MOBILE} {
      margin-top: ${MOBILE_MARGIN};
      margin-left: ${MOBILE_MARGIN};
    }
  `
      : `
    position: fixed;
    z-index: ${Z_INDEXES.NOTIFICATION};
    ${top !== undefined ? `top: ${top}px;` : ''}
    ${left !== undefined ? `left: ${left}px;` : ''}
    ${right !== undefined ? `right: ${right}px;` : ''}
    &:not(:first-child) {
      top: ${TOP_MARGIN};
    }
    @media ${MEDIA_MOBILE} {
      top: ${MOBILE_MARGIN};
      left: ${MOBILE_MARGIN};
    }
  `}

  @media ${MEDIA_MOBILE} {
    width: calc(100% - 2 * ${MOBILE_MARGIN});
    padding: ${MOBILE_DYNAMIC_PADDING};

    ${Icon} {
      display: none;
    }

    ${({ mobile }) =>
      mobile &&
      css`
        ${IconClose} {
          display: none;
        }
      `}
  }

  ${({ theme, kind, inverse }) =>
    inverse
      ? `
        background-color: ${theme.color.neutral[80]};
        border-left: 4px solid ${getInverseColor(kind, theme)};
        ${Header}, ${Message} {
          color: ${theme.color.neutral.white};
        }
        ${Icon} svg {
          fill: ${getInverseColor(kind, theme)};
        }
        ${IconClose} svg {
          fill: ${theme.color.neutral.white};
          &:hover {
            fill: ${theme.color.primary[40]};
          }
        }`
      : `
        box-shadow: ${theme.boxShadow[4]};
        ${IconClose} svg {
          fill: ${theme.color.neutral[50]};
          &:hover {
            fill: ${theme.color.primary[60]};
          }
        }
  `}
`;

DynamicNotificationComponent.defaultProps = {
  theme: DEFAULT_THEME,
};
