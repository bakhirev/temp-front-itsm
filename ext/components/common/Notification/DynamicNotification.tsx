import React, { useState, useEffect, FC } from 'react';
import { Transition } from 'react-transition-group';
import ReactDOM from 'react-dom';

import { ReactComponent as CloseOutline } from '../../Icons/service/CloseOutline.svg';
import { useGesture } from '../hooks';
import { useMobile } from '../hooks/useMobile';

import type { INotificationProps } from './NotificationComponent';
import { NOTIFICATION_ICONS } from './NotificationComponent';
import { DynamicNotificationComponent } from './DynamicNotificationComponent';
import { Icon } from './Icon';
import { IconClose } from './IconClose';
import { DEFAULT_DURATION } from './constants';
import { getTransitionStyles } from './utils';

export type VisibilityHandler = () => void;

export interface IDynamicNotificationProps extends INotificationProps {
  /** Время отображения анимации открытия/закрытия */
  duration?: number;
  /** Время отображения оповещения до автоматического закрытия */
  delay?: number;
  /** Контейнер, в котором происходит размещение оповещения */
  container?: Element;
  /** Коллбэк начала анимации отображения оповещения */
  onShow?: VisibilityHandler;
  /** Коллбэк окончания анимации скрытия оповещения */
  onHide?: VisibilityHandler;
}

const DEFAULT_WIDTH = '100%';

interface IDynamicNotification extends IDynamicNotificationProps {
  inverse?: boolean;
  top?: number;
  left?: number;
  right?: number;
}

export const DynamicNotification: FC<IDynamicNotification> = ({
  children,
  kind = 'info',
  width = DEFAULT_WIDTH,
  duration = DEFAULT_DURATION,
  delay,
  container,
  header,
  inverse,
  top,
  left,
  right,
  className,
  dataTestId,
  onShow,
  onHide,
  ...props
}) => {
  const KindIcon = NOTIFICATION_ICONS[kind];
  const mobile = useMobile();
  const [gesture, gestureBinding] = useGesture();
  const [visible, setVisible] = useState(true);
  const [isMounted, setIsMounted] = useState(false);
  useEffect(() => {
    setIsMounted(true);
  }, []);

  useEffect(() => {
    (gesture === 'left' || gesture === 'right') && setVisible(false);
  }, [gesture, setVisible]);

  const applyTimer = () => {
    if (delay) {
      const timer = setTimeout(() => {
        setVisible(false);
        clearTimeout(timer);
      }, delay);
    }
  };

  const transitionStyles = getTransitionStyles(duration);
  const component = (
    <Transition
      in={isMounted && visible}
      timeout={duration}
      onEntered={() => applyTimer()}
      onEnter={() => onShow && onShow()}
      onExited={() => onHide && onHide()}
    >
      {(state) => (
        <DynamicNotificationComponent
          {...gestureBinding}
          {...props}
          mobile={mobile}
          style={{ ...transitionStyles[state] }}
          className={className}
          data-test-id={dataTestId}
          kind={kind}
          width={width}
          inverse={inverse}
          top={top}
          left={left}
          right={right}
        >
          <Icon>
            <KindIcon />
          </Icon>
          <IconClose onClick={() => setVisible(false)} inverse={inverse} markerOffset={6}>
            <CloseOutline width={20} height={20} />
          </IconClose>
          {children}
        </DynamicNotificationComponent>
      )}
    </Transition>
  );
  return container ? ReactDOM.createPortal(component, container) : component;
};
