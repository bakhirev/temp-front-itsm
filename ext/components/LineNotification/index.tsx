import React, { FC } from 'react';

import { Header } from '../common/Notification/Header';
import { Message } from '../common/Notification/Message';
import { DynamicNotification } from '../common/Notification/DynamicNotification';
import type { IDynamicNotificationProps } from '../common/Notification/DynamicNotification';

export interface ILineNotificationProps extends IDynamicNotificationProps {
  /** Принудительно тёмный фон */
  showDarkBackground?: boolean;
}

export const LineNotification: FC<ILineNotificationProps> = ({
  children,
  header,
  showDarkBackground,
  ...props
}) => (
  <DynamicNotification inverse={showDarkBackground} top={0} left={0} {...props}>
    {!!header && <Header>{header}</Header>}
    <Message>{children}</Message>
  </DynamicNotification>
);
