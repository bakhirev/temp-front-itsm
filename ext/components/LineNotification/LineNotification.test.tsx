import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';

import { runCommonTests, runCommonDynamicTests } from '../common/Notification/test';
import { LineNotification } from '../LineNotification';

const message = 'LineNotification Message';

jest.mock('../common/default-theme', () => ({
  DEFAULT_THEME: {},
}));

describe('LineNotification', () => {
  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((elementTransition) => {
      return elementTransition && elementTransition.props && elementTransition.props.children();
    }) as any;
  });

  afterEach(() => {
    (ReactDOM.createPortal as any).mockClear();
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  runCommonTests(LineNotification, 'LineNotification');
  runCommonDynamicTests(LineNotification, 'LineNotification');

  it('should render component by default as "INFO" notification with dark background', () => {
    const wrapper = shallow(<LineNotification showDarkBackground>{message}</LineNotification>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind="ERROR" notification with dark background', () => {
    const wrapper = shallow(
      <LineNotification kind={'error'} showDarkBackground>
        {message}
      </LineNotification>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
