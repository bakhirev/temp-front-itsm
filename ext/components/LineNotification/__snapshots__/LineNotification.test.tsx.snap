// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineNotification should render component by default as "INFO" notification with dark background 1`] = `
<Component
  inverse={true}
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component by default as "INFO" notification with header 1`] = `
<Component
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Header
  </Styled(styled.div)>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component by default as "INFO" notification without header 1`] = `
<Component
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component by default as "INFO" notification without header with custom width 1`] = `
<Component
  left={0}
  top={0}
  width={777}
>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component inside the document.body container 1`] = `
<Component
  container={<body />}
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component with kind="ERROR" notification with dark background 1`] = `
<Component
  inverse={true}
  kind="error"
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component with kind="ERROR" notification with header 1`] = `
<Component
  kind="error"
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Header
  </Styled(styled.div)>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component with kind="ERROR" notification without header 1`] = `
<Component
  kind="error"
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component with kind="INFO" notification with header 1`] = `
<Component
  kind="info"
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Header
  </Styled(styled.div)>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component with kind="INFO" notification without header 1`] = `
<Component
  kind="info"
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component with kind="SUCCESS" notification with header 1`] = `
<Component
  kind="success"
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Header
  </Styled(styled.div)>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component with kind="SUCCESS" notification without header 1`] = `
<Component
  kind="success"
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component with kind="WARNING" notification with header 1`] = `
<Component
  kind="warning"
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Header
  </Styled(styled.div)>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;

exports[`LineNotification should render component with kind="WARNING" notification without header 1`] = `
<Component
  kind="warning"
  left={0}
  top={0}
>
  <Styled(styled.div)>
    LineNotification Message
  </Styled(styled.div)>
</Component>
`;
